src/fs/base/tree.ts

```diff
@@ -52,105 +42,121 @@ abstract class BaseTree implements Tree, UnixTree {
+      })
     }
-
-    const toAddNode = check.isTree(toAdd) ? toAdd : await this.createChildFile(child as FileContent, name)
-
-    return this.updateDirectChild(toAddNode, name)
+    return this
   }
 
   async rm(path: string): Promise<this> {
-    const parts = pathUtil.splitNonEmpty(path)
-    if (parts === null) {
-      throw new Error("Path does not exist")
-    }
-    await this.rmNested(parts)
-    await this.put()
+    await this.rmRecurse(path, () => this.put())
     return this
   }
 
-  async rmNested(path: NonEmptyPath): Promise<this> {
-    const filename = path[path.length - 1]
-    const parentPath = path.slice(0, path.length - 1)
-    const node = await this.get(pathUtil.join(parentPath))
-
-    if (node === null || check.isFile(node)) {
-      throw new Error("Path does not exist")
+  async rmRecurse(path: string, onUpdate: Maybe<UpdateCallback>): Promise<this> {
+    const { head, nextPath } = pathUtil.takeHead(path)
+    if(head === null){
+      throw new Error("Invalid path: empty")
     }
-
-    const updated = await node.removeDirectChild(filename)
-    return parentPath.length > 0
-            ? this.addChild(pathUtil.join(parentPath), updated)
-            : updated as this
+    if(nextPath === null) {
+      this.removeDirectChild(head)
+      onUpdate && await onUpdate()
+    }else {
+      const child = await this.getDirectChild(head)
+      if(child === null) {
+        throw new Error("Invalid path: does not exist")
+      } else if(check.isFile(child)) {
+        throw new Error(`There is a file along the given path: ${path}`)
+      } 
+      await child.rmRecurse(nextPath, async () => {
+        await this.updateDirectChild(child, head, onUpdate)
+      })
+    }
+    return this
   }
 
   async mv(from: string, to: string): Promise<this> {
-    const node = await this.get(from)
-    if (node === null) {
-      throw new Error(`Path does not exist: ${from}`)
-    }
-    const toParts = pathUtil.splitParts(to)
-    const destPath = pathUtil.join(toParts.slice(0, toParts.length - 1)) // remove file/dir name
-    const destination = await this.get(destPath)
-    if (check.isFile(destination)) {
-      throw new Error(`Can not \`mv\` to a file: ${destPath}`)
-    }
-
-    check.isTree(node)
-      ? await this.addChild(to, node)
-      : await this.addChild(to, node.content)
-
-    return this.rm(from)
+    throw new Error("mv has been temporarily disabled")
+    // const node = await this.get(from)
+    // if (node === null) {
+    //   throw new Error(`Path does not exist: ${from}`)
+    // }
+    // const { tail, parentPath } = pathUtil.takeTail(to)
+    // if(tail === null){
+    //   throw new Error(`Path does not exist: ${to}`)
+    // }
+    // const parent = await this.get(parentPath || '')
+    // if (check.isFile(parent)) {
+    //   throw new Error(`Can not \`mv\` to a file: ${parentPath || ''}`)
+    // }
+
+    // await this.updateDirectChild(node, tail, null)
+
+    // return this.rm(from)
```

Comment:
Why disabled temporarily? Ahead of versioning stuff? WIP?

Reply:
Found a bug with moving private stuff. Daniel said he'd fix it after these 2 PRs.

Reply:
yup^^
> I've realized that it does not work properly on the private side due to 1) lack of signposting & 2) not adjusting the namefilter on move
> Both of these can be fixed, but I'm thinking I might do that after I get concurrency & versioning/events working

I figured it was more important to get these changes out & start testing them. & signposting is going to be related to the events stuff anyway, so I figured I'd implement it then